File: backend/app/routes/shops.py
from fastapi import APIRouter, Depends, HTTPException
from bson import ObjectId
from ..db import db
from ..auth import role_required
from ..utils.qrcode import generate_shop_qr

router = APIRouter()


@router.post('/', dependencies=[Depends(role_required(['admin']))])
async def create_shop(data: dict):
    owner_id = data.get('ownerId')
    if not owner_id:
        raise HTTPException(status_code=400, detail='ownerId required')
    owner = await db.users.find_one({'_id': ObjectId(owner_id)})
    if not owner:
        raise HTTPException(status_code=400, detail='Owner not found')
    shop = {'name': data['name'], 'slug': data['slug'], 'owner': owner['_id']}
    res = await db.shops.insert_one(shop)
    await db.users.update_one({'_id': owner['_id']}, {'$set': {'shop': res.inserted_id, 'role': 'shopOwner'}})
    qr = generate_shop_qr(shop['slug'])
    shop['id'] = str(res.inserted_id)
    return {'shop': shop, 'qr': qr}


@router.get('/{slug}/qr')
async def get_qr(slug: str):
    shop = await db.shops.find_one({'slug': slug})
    if not shop:
        raise HTTPException(status_code=404, detail='Not found')
    qr = generate_shop_qr(shop['slug'])
    return {'qr': qr}

File: backend/src/routes/shops.js
import express from 'express';
import { roleMiddleware } from '../middleware/auth.js';
import Shop from '../models/Shop.js';
import User from '../models/User.js';
import { generateShopQr } from '../utils/qrcode.js';

const router = express.Router();

router.post('/', roleMiddleware(['admin']), async (req, res) => {
  const shop = await Shop.create({ ...req.body, owner: req.body.ownerId });
  await User.findByIdAndUpdate(req.body.ownerId, { shop: shop.id, role: 'shopOwner' });
  const qr = await generateShopQr(shop.slug);
  res.json({ shop, qr });
});

router.get('/:slug/qr', async (req, res) => {
  const shop = await Shop.findOne({ slug: req.params.slug });
  if (!shop) return res.status(404).json({ message: 'Not found' });
  const qr = await generateShopQr(shop.slug);
  return res.json({ qr });
});

export default router;
